### PR TITLE
Draft missing from post list

### PR DESCRIPTION
**Pull request by elibud**
**Fixes** #8916 

**To test:**

1. Go to the posts list
2. Display the drafts list
3. Open the editor to create a new Post
4. Save it as a Draft from the top right three dots menu
5. Close the editor from the cross on the top left and to go back to the Draft list
6. The newly created post should be there

Also test: publishing a new post from the draft list, from the published list, scheduling posts, etc. Basically we need to run through all the publishing scenarios to make sure nothing is broken.

Notes:

There were two issues at play here:
 - The `shouldRemovePostOnDismiss` property was being calculated during the ViewController initialization, which was incorrect because the post status could change later on.
- When closing the editor from the top left cross, and assumption that there were no changes was made, and this is not correct since the post could've been previously saved via the right three dots menu. Basically when we close the editor we can't assume no changes were made unless the post was created and deleted/discarded.

I've made the minimum amount of changes that fix the bug, however there's a lot of dead code on `PostEditor+Publish`, and logic around PostPost that should be cleaned up in a subsequent PR.

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by elibud** [comment]
Thanks @pinarol and I agree, we definitely want more eyes on it. I'll wait until either @diegoreymendez or @koke take it for a spin.

**Comment by diegoreymendez** [comment]
Fwiw, I think this code is complex enough that no-one can really give a good answer.  I'm wrapping something up but will give it a look as soon as possible.

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift [hunk 1]
```diff
@@ -374,13 +373,13 @@ extension PostEditor where Self: UIViewController {
...
 
-    func dismissOrPopView(didSave: Bool, shouldShowPostEpilogue: Bool = true) {
+    func dismissOrPopView() {
         stopEditing()
 
         WPAppAnalytics.track(.editorClosed, withProperties: [WPAppAnalyticsKeyEditorSource: analyticsEditorSource], with: post)
 
         if let onClose = onClose {
-            onClose(didSave, shouldShowPostEpilogue)
+            onClose(true, false)
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I'm resisting the urge to suggest further refactors, since this is so hard to follow, but if I had to pick just one, this method could really benefit from some argument labels.

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift [hunk 2]
```diff
@@ -232,7 +231,7 @@ extension PostEditor where Self: UIViewController {
         post = originalPost
         post.deleteRevision()
 
-        if shouldRemovePostOnDismiss {
+        if post.hasNeverAttemptedToUpload() && !post.isLocalRevision {
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I think `shouldRemovePostOnDismiss` was more descriptive of what was being checked, perhaps we can keep it as a computed property?

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift (hunk 1 above)
Comment (elibud) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: Yes, I felt the same way while writing the code, there are a ton of improvements that could be made here, but on this PR I'm just trying to fix the bug.

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift (hunk 2 above)
Comment (elibud) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I'm reluctant to have a property for something that's calculated only once and is a one liner &&. I also want to avoid having to lose the context of the method when you go check what `shouldRemovePostOnDismiss` really means.
But if you feel strongly about it I'll change it :D

**Comment by elibud** [comment]
@koke Yes, changing previous behavior was the only way to fix this one. Basically you can't really know if changes were saved or not when you close the editor from the top left cross, any value for the `changesSaved` parameter will be incorrect on some cases. If you always send `false` as we were doing it would be incorrect for any post saved from the three dot menu.
Good find on the testing, I'm checking to see if I can figure out a solution.

**Comment by elibud** [comment]
Ready for another look, now instead of assuming changes were always saved, I'm assuming they were always saved unless the post was deleted.

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift (hunk 2 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I agree with both (so helpful, I know!).

One middle ground option (that I often use when I want to provide more clarity) could be something along the lines of:

```swift
let shouldRemovePostOnDismiss =
        post.hasNeverAttemptedToUpload()
        && !post.isLocalRevision

if shouldRemovePostOnDismiss {
```

It doesn't really change the logic, but makes the code slightly more readable.

Comment (elibud) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I like it, changes made!

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift [hunk 3]
```diff
@@ -192,19 +192,20 @@ extension PostEditor where Self: UIViewController {
...
+            discardUnsavedChangesAndUpdateGUI()
         }
     }
 
-    func discardChangesAndUpdateGUI() {
-        discardChanges()
-
-        dismissOrPopView(didSave: false)
+    func discardUnsavedChangesAndUpdateGUI() {
+        let postDeleted = discardChanges()
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I wasn't supper happy about this, since it's not obvious that a method named `discardChanges` would return whether a post was deleted or not, and then `didSave == !postDeleted` seems like a non-obvious assumption. However I kept following the code and the whole thing is quite confusing, so I'll 🙈 for now and hope that having autosaves eventually allows us to simplify all this logic.

Comment (elibud) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I didn't like that either, but yes, the whole thing is a mess, we need to rewrite it.

Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift]: I agree.  I think it would be nice to try to think of ways in which we can incrementally improve this.
